### ISSUE TITLE
 saving a Cyclops model to json doesn't save the names of the coefficients

### ISSUE THREAD

**Issue by egillax** [OHDSI/PatientLevelPrediction#305]
In the current develop branch saving a Cyclops model to json doesn't save the names of the coefficients. This causes external validation to fail because in ```predictCyclops``` it needs the names to extract the coefficients. 

This seems to be because we are now using ```ParallelLogger::saveSettingsToJson``` to save the model instead of ```jsonlite::serializeJson```. This doesn't preserve names of named vectors like the Cyclops coefficients. 

I'm not sure what would be the best way to resolve this, I assume @jreps you changed the saving because serializeJson produces pretty verbose and complicated JSON's ? Maybe we could use a list instead of a numeric vector for the coefficients? But that would probably require many small changes throughout the code.

**Comment by egillax** [comment]
A simple way of reproducing the issue:

```R
connectionDetails <- Eunomia::getEunomiaConnectionDetails()
Eunomia::createCohorts(connectionDetails)

covSet <- FeatureExtraction::createCovariateSettings(useDemographicsGender = T,
                                                     useDemographicsAge = T,
                                                     useDemographicsRace = T,
                                                     useDemographicsEthnicity = T,
                                                     useDemographicsAgeGroup = T,
                                                     useConditionGroupEraLongTerm = T,
                                                     useDrugEraStartLongTerm  = T,
                                                     endDays = -1
)

databaseDetails <- PatientLevelPrediction::createDatabaseDetails(
  connectionDetails = connectionDetails,
  cdmDatabaseSchema = "main",
  cohortDatabaseSchema = "main",
  cohortTable = "cohort",
  targetId = 4,
  outcomeIds = 3,
  outcomeDatabaseSchema = "main",
  outcomeTable =  "cohort",
  cdmDatabaseName = 'eunomia'
)

restrictPlpDataSettings <- PatientLevelPrediction::createRestrictPlpDataSettings(
  firstExposureOnly = T,
  washoutPeriod = 365
)
plpData <- PatientLevelPrediction::getPlpData(
  databaseDetails = databaseDetails,
  restrictPlpDataSettings = restrictPlpDataSettings,
  covariateSettings = covSet
)


populationSettings <- PatientLevelPrediction::createStudyPopulationSettings(
  requireTimeAtRisk = F,
  riskWindowStart = 1,
  riskWindowEnd = 365)

saveLoc <- tempdir()
modelSettings <- PatientLevelPrediction::setLassoLogisticRegression()
plpResults <- PatientLevelPrediction::runPlp(plpData = plpData,
                                             outcomeId = 3,
                                             modelSettings = modelSettings,
                                             analysisId = 'Test',
                                             analysisName = 'Testing DeepPlp',
                                             populationSettings = populationSettings,
                                             splitSettings = createDefaultSplitSetting(),
                                             sampleSettings = createSampleSettings(),  # none
                                             featureEngineeringSettings = createFeatureEngineeringSettings(), # none
                                             preprocessSettings = createPreprocessSettings(),
                                             logSettings = createLogSettings(),
                                             executeSettings = createExecuteSettings(runSplitData = T,
                                                                                     runSampleData = T,
                                                                                     runfeatureEngineering = T,
                                                                                     runPreprocessData = T,
                                                                                     runModelDevelopment = T,
                                                                                     runCovariateSummary = T
                                             ),
                                             saveDirectory = saveLoc)

plpModel <- loadPlpModel(file.path(saveLoc, 'Test', 'plpResult', 'model'))

population <- createStudyPopulation(plpData, outcomeId=3, populationSettings = populationSettings)
predictPlp(plpModel=plpModel, plpData, population = population)

```

This gives the warning 

```Warning message:
In predictCyclopsType(plpModel$model$coefficients, cohort, data$covariateData,  :
  Model had no non-zero coefficients so predicted same for all population...```

**Comment by jreps** [comment]
Wow - I did not realize it dropped the names in the vector.  I remember there was some issue with ParallelLogger that impacted saving/loading the json but I thought it was fixed.

We need to make this an issue in ParallelLogger to get it resolved but for the time being we need some temporary fix.

The saving to json has been a bit of a nightmare - I found out the sklearn-json package is now outdated so we cannot save the python models to json anymore (unless we update the sklearn-json package ourselves).

**Comment by jreps** [comment]
I think for a fix I can change the coefficient object to be a data.frame with the covariateids and betas - this is currently done during the prediction, but I can move this to where the model is created instead

**Comment by egillax** [comment]
I think that fix would be good. I'd also suggest changing the test for external validation.  Currently in this [line](https://github.com/OHDSI/PatientLevelPrediction/blob/develop/tests/testthat/test-validation.R#L36) it's loading the model from memory:

```R 
modelVal <- plpResult$model 
```

But the actual behavior when externally validating is loading the model from a file because someone sends you the model. So loading it in the test from a file would better reflect the actual behavior, and that would also have caught this bug.

I'd be happy to submit a PR for this and also the proposed fix.

**Comment by egillax** [comment]
Fixed in #307
